src/mpi_t/events_impl.c

```diff
@@ -178,25 +206,75 @@ int MPIR_T_event_get_timestamp_impl(MPI_T_event_instance event_instance,
 int MPIR_T_event_handle_alloc_impl(int event_index, void *obj_handle, MPIR_Info * info_ptr,
                                    MPI_T_event_registration * event_registration)
 {
-    return MPI_T_ERR_INVALID_INDEX;
+    MPIR_T_event_t *event;
+    HASH_FIND_INT(events, &event_index, event);
+    if (event == NULL) {
+        return MPI_T_ERR_INVALID_INDEX;
+    }
+
+    MPIR_T_event_registration_t *event_reg =
+        MPL_malloc(sizeof(MPIR_T_event_registration_t), MPL_MEM_MPIT);
+#ifdef HAVE_ERROR_CHECKING
+    event_reg->kind = MPIR_T_EVENT_REG;
+#endif
+    event_reg->event = event;
+    LL_APPEND(event->reg_list_head, event->reg_list_tail, event_reg);
+    if (event->bind != MPI_T_BIND_NO_OBJECT)
+        event_reg->obj_handle = obj_handle;
+
+    /* initialize callbacks */
+    memset(event_reg->callbacks, 0, sizeof(event_reg->callbacks));
+    event_reg->dropped_cb = NULL;
+    event_reg->dropped_count = 0;
+
+    *event_registration = event_reg;
+
+    return MPI_SUCCESS;
 }
 
 int MPIR_T_event_handle_free_impl(MPI_T_event_registration event_registration, void *user_data,
                                   MPI_T_event_free_cb_function free_cb_function)
 {
-    return MPI_T_ERR_INVALID_HANDLE;
+#ifdef HAVE_ERROR_CHECKING
+    if (event_registration->kind != MPIR_T_EVENT_REG) {
+        return MPI_T_ERR_INVALID_HANDLE;
+    }
+#endif
```

Comment:
Should we consider to amend python script to add this validation to the MPI_T_event_registration  kind?

Reply:
done